### ISSUE TITLE
Small Demo

### ISSUE THREAD

**Issue by reedkotler** [google/ml-compiler-opt#303]
What would be really helpful is a small test case that can train in 30 minutes on a modest machine. It does not have to produce a useful model but is something that one can see end to end without days of training. I'm willing to help make the test case if I can get some help.

**Comment by mtrofin** [comment]
Yup, and we could even set it up in the CI as a nightly. Do you have a project in mind to play "corpus" - if not, llvm itself could be it (if it wouldn't make things more confusing by playing 2 roles)

**Comment by mtrofin** [comment]
Right, so we could use llvm itself as the corpus donor. Sticking to inlining for size - because regalloc would need profiles (and I think for end to end demo-ing, inline for size is a fine example). These are the steps:

* git clone llvm to, for instance, `/work/llvm-project`, `cd /work/llvm-project`, `mkdir tflite-build && cd tflite-build`. cmake will need to additionally have `-DLLVM_ENABLE_PROJECTS=clang`. The goal of this step is to build the clang we'll use for training, but we'll also use this clang for corpus collection. `ninja clang llvm-objcopy` (we need objcopy to extract the corpus)
* `cd .. && mkdir corpus-build && cd corpus-build`
* `cmake -GNinja -DCMAKE_BUILD_TYPE=MinSizeRel -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_CXX_COMPILER=/work/llvm-project/build/bin/clang++ -DCMAKE_C_COMPILER=/work/llvm-project/build/bin/clang -DCMAKE_CXX_FLAGS="-Xclang=-fembed-bitcode=all" -DCMAKE_C_FLAGS="-Xclang=-fembed-bitcode=all" ../llvm`. We don't bother trying to say we build for size, the goal is just to get to a corpus. Note this generates a `compile_commands.json` in that build dir.
* `ninja opt llc` -> this is so we have some objects built.
* assuming you git clone-d ml-compiler-opt under /work/ml-compiler-opt: `cd /work/ml-compiler-opt` then `PYTHONPATH=$PYTHONPATH:. python3 compiler_opt/tools/extract_ir.py --input /work/llvm-project/corpus-build/compile_commands.json --input_type json --llvm_objcopy_path /work/llvm-project/build/bin/llvm-objcopy --output_dir /tmp/corpus`
* This extracts the corpus in `/tmp/corpus`, after that the training steps are the same as in the demos - i.e. collect a default trace.. all that.

**Comment by reedkotler** [comment]
thanks. i'll try these steps out.
________________________________

**Comment by pshung** [comment]
What training performance could I expect if only using LLVM as the training corpus?

**Comment by mtrofin** [comment]
What do you mean by "training performance": time it takes to train a model? Or model effectiveness (i.e. how much that model can shrink binaries)? Either way, I think @reedkotler did this recently (llvm as corpus), perhaps he can comment on both.

Fuchsia's case discussed in the demo used to be half a day, but when we doubled the feature count, so did the training time. IIRC they get ~3% shrinkage in their overall image.

**Comment by pshung** [comment]
Thanks for your instruction about using LLVM as a training corpus. I was able to run the inlining training.
However, LLVM corpus includes about 2800 modules only.
so, I wonder if the size reduction and the generalization ability against the performance figure mentioned in the paper (It claims 28000 IR modules to reach that performance).

**Comment by mtrofin** [comment]
In general, we observed that having more modules from more diverse projects, during training, would help a model generalize better, but just like with manual heuristics, without trying it out, it's hard to tell what to expect for a specific case.

**Comment by ioana-ghiban-arm** [comment]
Hello! 

What would be the steps for 'Deploying and using the new policy' when using LLVM itself as the corpus donor? I have _hopefully_ trained the optimized model from the warmstart model, but in $OUTPUT_DIR I only see `policy` dir, no `saved_policy`, not sure if that can be the reason for which I can't [build the release](https://github.com/google/ml-compiler-opt/blob/main/docs/inlining-demo/demo.md#:~:text=We%20need%20to%20build%20the%20%27release%27%20mode%20of%20the%20compiler.%20Currently%2C%20that%20means%20overwritting%20the%20model%20in%20llvm/lib/Analysis/models/inliner.). I configured it this way:
`cmake -G Ninja \
-DCMAKE_BUILD_TYPE=Release \
-DLLVM_ENABLE_PROJECTS="clang" \
-DTENSORFLOW_AOT_PATH=${TENSORFLOW_AOT_PATH} \
-C $TFLITE_PATH/tflite.cmake \
$LLVM_SRCDIR/llvm`
Another deviation from the demo instructions in my setup is that I hardcoded $TENSORFLOW_AOT_PATH to be able to generate the ninja. 

Any suggestions would be much appreciated. Thanks.

**Comment by ioana-ghiban-arm** [comment]
Indeed, `train_locally.py` didn't run for long enough. The time it usually takes to create `saved_policy` seems to be around 10 hrs. However, I see the same performance (no binary size change) with this final 'saved_policy` as I did with an intermediary one from some folder in `policy`. As mentioned [here](https://github.com/google/ml-compiler-opt/issues/379#issuecomment-2401652632), in the case of using clang for both training and corpus collection, I've only noticed a difference in the size of the `llvm-config` binary, marginally larger when created by clang built with `-DLLVM_INLINER_MODEL_PATH=/path/to/model/saved_policy`. 
Do you have any suggestions for tweaking my setup? I could share the steps I've taken but that might get too verbose for an issue comment. I've basically filled in the steps I was missing from the directions above with what seemed most sensible in either the inlining or regalloc demos.

**Comment by mtrofin** [comment]
OK, that's weird. Let's first make sure the use side of things - i.e. how the model is ingested and used - is set up right. Then we can look at the training side.

I tried the published size model, here are my exact steps:


For brevity, I used my paths - I have a git repo for llvm under `/work/llvm-project` and I have a python 3.10 env set up under `/work/python3.10`, so those paths need replacing. The toolchain used to bootstrap clang shouldn't matter.


```
cd /tmp
wget https://github.com/google/ml-compiler-opt/releases/download/inlining-Oz-v1.1/inlining-Oz-99f0063-v1.1.tar.gz
tar xvfz inlining-Oz-99f0063-v1.1.tar.gz
ls /tmp/model
```

```
cd /work/llvm-project
git checkout main && git pull && git checkout 665457815f11118f7e755a471f33606c8562a4be
```

```
mkdir build && cd build
cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm  -DLLVM_ENABLE_PROJECTS=clang  -DTENSORFLOW_AOT_PATH=/work/python3.10/lib/python3.10/site-packages/tensorflow -DLLVM_INLINER_MODEL_PATH=/tmp/model
ninja clang
```

```
cd ../ && mkdir build-base && cd build-base
cmake -GNinja -DCMAKE_BUILD_TYPE=MinSizeRel ../llvm  -DLLVM_ENABLE_PROJECTS=clang  -DTENSORFLOW_AOT_PATH=/work/python3.10/lib/python3.10/site-packages/tensorflow -DLLVM_INLINER_MODEL_PATH=/tmp/model -DCMAKE_C_COMPILER=/work/llvm-project/build/bin/clang -DCMAKE_CXX_COMPILER=/work/llvm-project/build/bin/clang++ -DCMAKE_EXPORT_COMPILE_COMMANDS=On
```

Check `compile-commands.json` to make sure it's using the previously built clang, then `ninja clang` (well, just `ninja` I guess - but I only built clang for validation)

```
cd ../ && mkdir build-exp && cd build-exp
cmake -GNinja -DCMAKE_BUILD_TYPE=MinSizeRel ../llvm  -DLLVM_ENABLE_PROJECTS=clang  -DTENSORFLOW_AOT_PATH=/work/python3.10/lib/python3.10/site-packages/tensorflow -DLLVM_INLINER_MODEL_PATH=/tmp/model -DCMAKE_C_COMPILER=/work/llvm-project/build/bin/clang -DCMAKE_CXX_COMPILER=/work/llvm-project/build/bin/clang++ -DCMAKE_EXPORT_COMPILE_COMMANDS=On -DCMAKE_C_FLAGS="-mllvm -enable-ml-inliner=release" -DCMAKE_CXX_FLAGS="-mllvm -enable-ml-inliner=release"
```

I'd check again `compile-commands.json` to make sure it's using the previously built clang and that the flags are right (i.e. we compile with `-mllvm -enable-ml-inliner=release`). `ninja clang` again.

```
# mtrofin@mtrofin.c.googlers.com in /work/llvm-project on git:main o [8:05:40]
$ ls -l build-base/bin/clang-20
-rwxr-x--- 1 mtrofin primarygroup 179609320 Oct  9 07:57 build-base/bin/clang-20

$ ls -l build-exp/bin/clang-20
-rwxr-x--- 1 mtrofin primarygroup 158697880 Oct  9 08:05 build-exp/bin/clang-20

[8:07:54]
```

so that's about 12% size savings.

A possible gotcha: are you passing -mllvm -enable-ml-inliner=release? Are you building -Os or -Oz? (this latter is a weaker problem, the former is critical)
